Add blog post for industrial data to IIoT (#1111)
flowforge/website: src/blog/2023/09/making-sense-of-your-industrial-data.md

Author: once we settle on a title, I will offer up a proposal based on that.

---
title: Making sense of your industrial data
subtitle: A look at legacy industrial protocol data and making sense of it in an IIoT world.
description: A look at legacy industrial protocol data and making sense of it in an IIoT world.

> date: 2023-09-08

Reviewer: Not too certain myself, but wanted to come up with a proposal. Maybe you have a better idea than these.

> title: Making sense of your industrial data
> subtitle: A look at legacy industrial protocol data and making sense of it in an IIoT world.
> description: A look at legacy industrial protocol data and making sense of it in an IIoT world.
> date: 2023-09-08

Author: Counter proposal.  Because:
1. Modbus is not (really) itself legacy, just that many legacy devices _only_ have one industrial protocol (non (I)IoT) and that protocol is quite often Modbus. (dont want to get in a flame war)
2. The article might get pigeon holed as only being for Modbus and thus 
   1. the graphic should be changed
   2. we lose traffic for those seeking a solution to {name_your_legacy_protocol} types of data
3. The title you propose suggests this is about simply accessing data. Many integrators can do that today with (expensive) OPC packages or HMIs.  The title I was going for was more about "normalisation" of "any" legacy data for an IIoT world.  I.e. not just Modbus but everyone (Serial comms with no protocol, Siemens S7, Omron Fins, Misti MC, Allen Bradley DF1, etc) should be aware there is a solution and we hold the key to it!

```suggestion
title: Modernize your legacy industrial data
subtitle: Working with legacy industrial protocol data from the likes of Modbus and older, non IIoT protocols and putting it to work in an IIoT world.
description: Working with legacy industrial protocol data from the likes of Modbus and older, non IIoT protocols and putting it to work in an IIoT world.
date: 2023-09-11
```

> date: 2023-09-08

Author: outdated, marking as resolved

<!--more-->

### Legacy Industrial Data: ModBus

ModBus is one of the most widely used industrial protocols. Originally developed in the late 1970s, it has been a popular choice for industrial communication ever since. However, its data format can be challenging to work with in the context of modern IIoT applications. ModBus typically represents data in 16-bit unsigned registers, making it necessary to convert this data into more usable formats like Signed integer, Float, Signed and Unsigned Long, String, or even individual bits.

### A short primer on data types

Before we dive into how to make sense of ModBus data, let's take a quick look at some of the common data types we have to deal with.

#### 16-bit unsigned

> 16-bit unsigned data is an integer that can only be positive. It can represent values from 0 to 65535. For example, the number 12345 is represented as `0x3039` in hexadecimal or `0011000000111001` in binary.

Reviewer: Not sure why a hexadecimal is introduced here, I'd just explain you have 16 bits for the value, and it's 2^16 values, as 0 is included the maximum values is 65535 (or 2^16 - 1).

#### 16-bit signed

> 16-bit signed data is an integer that can be positive or negative. It can represent values from -32768 to 32767. For example, the number -12345 is represented as `0xCFC7` in hexadecimal or `1100111111000111` in binary.

Reviewer: So now we need a bit to denote positive or negative, and so the number of values we can represent needs to be stored in 15 bits. So the maximum value is 2^15 -1

Author: As previous comment, I would much prefer to ignore scientific notation and gently introduce the reader to everyday constants. If you dont mind, I would prefer to leave as is.

#### 32-bit data

32-bit data, like 16-bit data can mean many things. It could be signed or unsigned, or even a floating point number. For example, the number 12345 is represented as `0x00003039` in hexadecimal or `00000000000000000011000000111001` in binary. Typically, 32-bit data is represented as two 16-bit registers. Therefore, when dealing with 32-bit data, you need to combine two 16-bit registers to get the full value.

#### Endianness

> Endianness is a term used to describe the order in which bytes are stored in memory. There are two types of endianness: big-endian and little-endian. In big-endian, the most significant byte is stored first, while in little-endian, the least significant byte is stored first. For example, the number 12345 is represented as `0x3039` in big-endian and `0x3930` in little-endian.  In other words, the bytes might be stored in reverse order. This further complicates the process of converting ModBus data into more usable formats.

Reviewer: ```suggestion
Endianness is a term used to describe the order in which bytes are stored in memory. There are two types of endianness: big-endian and little-endian. In big-endian, the most significant bit is stored first, while in little-endian, the least significant bit is stored first. For example, the number 12345 is represented as `0x3039` in big-endian and `0x3930` in little-endian.  In other words, the bits might be stored in reverse order. This further complicates the process of converting ModBus data into more usable formats.
```

The endianness is on a bit level, not byte level. :)

Author: I am afraid 100% acurate - in these terms of protocols.

The bytes are actually swapped not the bit ordering.

From the PoV of a buffer (of bytes), the functions doeing the work behind the scenes is the NodeJS buffer: https://nodejs.org/api/buffer.html#bufreadint16beoffset - the BE denoted BigEndian (byte swapped), the LE is ...

Other readme...

Stack overflow
> In modern computing endianness only applies to ordering of bytes, not bits

Wikipedia 
> In [computing](https://en.wikipedia.org/wiki/Computing), endianness is the order or sequence of [bytes](https://en.wikipedia.org/wiki/Byte) of a [word](https://en.wikipedia.org/wiki/Word_(data_type)) of digital data in [computer memory](https://en.wikipedia.org/wiki/Computer_memory) or [data communication](https://en.wikipedia.org/wiki/Data_communication) which is identified by describing the impact of the "first" bytes, meaning at the smallest address or sent first. Endianness is primarily expressed as big-endian (BE) or little-endian (LE). A big-endian system stores the [most significant byte](https://en.wikipedia.org/wiki/Most_significant_byte) of a word at the smallest [memory address](https://en.wikipedia.org/wiki/Memory_address) and the [least significant byte](https://en.wikipedia.org/wiki/Least_significant_byte) at the largest.

Some other site
>BE stores the big-end first. When reading multiple bytes the first byte (or the lowest memory address) is the biggest - so it makes the most sense to people who read left to right.
>LE stores the little-end first. When reading multiple bytes the first byte (or the lowest memory address) is the littlest -  so it makes most sense to people who read right to left.

Reviewer: Thanks, I didn't google, but would've bet my house on it just a few minutes ago. Today I learned (again)

### Node-RED and node-red-contrib-buffer-parser to the rescue

Node-RED is an open-source flow-based development tool for visual programming. It's particularly well-suited for IIoT applications because of its versatility and extensive library of nodes. One such node, node-red-contrib-buffer-parser, provides a solution to the legacy data conversion challenge.

> This powerful Node-RED module allows you to parse a Buffer of bytes or an Array of integer data (which, by no coincidence, the popular module node-red-contrib-modbus outputs), and convert it into various data types. It can output pretty much any data type, including byte-swapped data, WORD swapped data, masked/shifted/scaled data, and even individual bits.

Reviewer: ```suggestion
This powerful Node-RED module allows you to parse a Buffer of bytes or an Array of integer data (which, by no coincidence, the popular module `node-red-contrib-modbus` outputs), and convert it into various data types. It can output pretty much any data type, including byte-swapped data, WORD swapped data, masked/shifted/scaled data, and even individual bits.
```

Here's a quick overview of how it works:

1. **Data Parsing**: Start by setting up a ModBus READ node in Node-RED to retrieve data from your industrial device. Then, use the buffer parser node to parse the ModBus data.

1. **Data Conversion**: With the buffer parser, you can easily convert the 16-bit unsigned data into more meaningful formats. Whether you need to translate it into Float, Long, String, or even extract specific bits, this tool makes the process straightforward.

1. **Publishing to MQTT, influxDB, a dashboard, an IIoT system**: Once your data is in a usable format, Node-RED enables you to publish it to many places. MQTT (Message Queuing Telemetry Transport), a popular protocol for IIoT communication is a perfect example. This makes your data accessible to other IIoT systems and applications for further analysis and action.

### Unlocking the Potential of Legacy Data

By leveraging Node-RED and buffer parser, you can bridge the gap between legacy industrial protocols and the IIoT world. This means you can extract valuable insights from your existing infrastructure without the need for costly hardware upgrades or replacements.

> In the era of the Industrial Internet of Things, making sense of your industrial data is no longer a daunting challenge. With the right tools approach, you can unlock the full potential of your legacy data and drive efficiency, productivity, and innovation in your industrial processes.  Yey!

Reviewer: ```suggestion
In the era of the Industrial Internet of Things, making sense of your industrial data is no longer a daunting challenge. With the right tools approach, you can unlock the full potential of your legacy data and drive efficiency, productivity, and innovation in your industrial processes. Yey!
```

### Learn More

We will be publishing follow-up blog posts with more details, best practices and examples on how to use Node-RED to make sense of your industrial data. In the meantime, you can learn more about these tools by visiting the following links:
* [Node-RED blog posts](https://flowfuse.com/blog/node-red/)
* [Node-RED videos](https://www.youtube.com/playlist?list=PLpcyqc7kNgp09XeRx_cae1fEIOloPqM1C)
* [Buffer Parser Node](https://flows.nodered.org/node/node-red-contrib-buffer-parser) 

> ### A teaser

Reviewer: I think this is more than a teaser, and thus propose to move this section to above the 'Learn more' section, and call it "3 quick examples" or so.

Author: No problemo. Will do that last (after other parts are resolved.

Author: resolved in [6958872](https://github.com/flowforge/website/pull/1111/commits/6958872243434546368aa0dccea78bf37412a0c9)

Here are 3 quick demonstrations of Node-RED and the buffer parser node in action:

> #### Example 1

Reviewer: ```suggestion
#### Example 1: ModBus to MQTT
```

Converting an array of 16-bit unsigned integers to String, Float and a scaled integer and passing them to an MQTT broker in 4 nodes!
![Legacy data to MQTT](images/industrial-legacy-data-to-mqtt.gif)

> #### Example 2

Reviewer: ```suggestion
#### Example 2: ModBus to InfluxDB
```

Converting an array of 16-bit unsigned integers to String, Float and a scaled integer for publishing to influxDB!
![Legacy data to influxDB](images/industrial-legacy-data-to-influx.png)
![Legacy data to influxDB2](images/industrial-legacy-data-to-influx2.png)

> #### Example 3

Reviewer: ```suggestion
#### Example 3: ModBus data on a dashboard
```

---
title: Modernize your legacy industrial data
subtitle: Working with legacy industrial protocol data from the likes of Modbus and older, non IIoT protocols and putting it to work in an IIoT world.
description: Working with legacy industrial protocol data from the likes of Modbus and older, non IIoT protocols and putting it to work in an IIoT world.
date: 2023-09-11
authors: ["stephen-mclaughlin"]
image: "/blog/2023/09/images/industrial-legacy-data-blog-image.png"
tags:
    - posts
    - node-red
    - community
---

> In today's rapidly evolving industrial landscape, the ability to collect, analyse, and act upon data is more critical than ever.
> The Industrial Internet of Things (IIoT) has ushered in a new era of connectivity and automation, but what about the wealth of data locked away in legacy industrial protocols like ModBus?
> How can we unlock and make sense of this valuable information in the IIoT era? That's where tools like Node-RED and `node-red-contrib-buffer-parser` come into play.

Reviewer: ```suggestion
In today's rapidly evolving industrial landscape, the ability to collect, analyse, and act upon data is more critical than ever.
The Industrial Internet of Things (IIoT) has ushered in a new era of connectivity and automation, but what about the wealth of data locked away in legacy industrial protocols like ModBus? How can we unlock and make sense of this valuable information in _any_ data, e.g. serial comms with no protocol, Siemens S7, Omron Fins, Misti MC, or Allen Bradley DF1 
in the IIoT era? That's where tools like Node-RED and `node-red-contrib-buffer-parser` come into play.
```

I liked your explanation on why the title is what it now is, so I copied part over to the intro. Let me know what you think.

Author: resolved in [6958872](https://github.com/flowforge/website/pull/1111/commits/6958872243434546368aa0dccea78bf37412a0c9)

### A short primer on data types

Before we dive into how to make sense of ModBus data, let's take a quick look at some of the common data types we have to deal with.

#### 16-bit unsigned

16-bit unsigned data is an integer that can only be positive. It can represent values from 0 to 65535. For example, the number 12345 is represented as `0x3039` in hexadecimal or `0011000000111001` in binary.

#### 16-bit signed

16-bit signed data is an integer that can be positive or negative. It can represent values from -32768 to 32767. For example, the number -12345 is represented as `0xCFC7` in hexadecimal or `1100111111000111` in binary.

#### 32-bit data

32-bit data, like 16-bit data can mean many things. It could be signed or unsigned, or even a floating point number. For example, the number 12345 is represented as `0x00003039` in hexadecimal or `00000000000000000011000000111001` in binary. Typically, 32-bit data is represented as two 16-bit registers. Therefore, when dealing with 32-bit data, you need to combine two 16-bit registers to get the full value.

#### Endianness

> Endianness is a term used to describe the order in which bytes are stored in memory. There are two types of endianness: big-endian and little-endian. In big-endian, the most significant byte is stored first, while in little-endian, the least significant byte is stored first. For example, the number 12345 is represented as `0x3039` in big-endian and `0x3930` in little-endian.  In other words, the bytes might be stored in reverse order. This further complicates the process of converting ModBus data into more usable formats.

Author: ```suggestion
Endianness, particularly in the context data communications, refers to the order or the sequence of how bytes are stored or transmitted. There are two types of endianness: big-endian (BE) and little-endian (LE). In big-endian, the most significant byte is first, while in little-endian, the least significant byte is first. For example, the number 12345 is represented as `0x3039` in a big-endian word and `0x3930` in little-endian word.  This can often cause confusion and complicate the process of converting ModBus data into more usable formats.
```

Perhaps this suggestion removes the confusion a little.  I dont want to fully remove confusion (😉) or even remove this section.  I feel it has value for demonstrating that "we" (Node-red/buffer parser) can help you make sense of the confusion and difficulties legacy protocols can present (and partly why the very next sentence says "Node-RED and node-red-contrib-buffer-parser to the rescue")
